File: python/tests/test_version.py
"""The repo carries one version, in VERSION.

Until v0.9.4 ``pyproject.toml`` and ``__init__.py`` both said ``0.1.0`` while
the repo shipped 0.9.3 -- nothing pinned them, unlike ``CONTRACT_VERSION``,
which :mod:`test_registry_sync` and :mod:`test_docs` both pin.
"""

from __future__ import annotations

import re

import pytest

import datalib


def _repo_version(repo_root) -> str:
    vfile = repo_root / "VERSION"
    if not vfile.is_file():
        pytest.skip("VERSION not found (running outside the repo)")
    return vfile.read_text(encoding="utf-8").strip()


def test_dunder_version_matches_repo(repo_root) -> None:
    assert datalib.__version__ == _repo_version(repo_root)


def test_pyproject_version_matches_repo(repo_root) -> None:
    text = (repo_root / "python" / "pyproject.toml").read_text(encoding="utf-8")
    m = re.search(r'^version\s*=\s*"([^"]+)"', text, re.MULTILINE)
    assert m, "no version key in pyproject.toml"
    assert m.group(1) == _repo_version(repo_root)

File: r/tests/testthat/test-version.R
# The repo carries one version, in VERSION. Until v0.9.4 r/DESCRIPTION said
# 0.1.0 and python/pyproject.toml said 0.1.0 while the repo shipped 0.9.3 --
# nothing pinned them, unlike CONTRACT_VERSION which two Python tests pin.

test_that("DESCRIPTION Version equals the repo VERSION file", {
  # datalib_test_repo_tests() returns <repo>/tests, so the repo is its parent
  repo <- dirname(datalib_test_repo_tests())
  vfile <- file.path(repo, "VERSION")
  skip_if_not(file.exists(vfile), "VERSION not found (running outside the repo)")
  want <- trimws(readLines(vfile, warn = FALSE)[[1]])
  # read.dcf keeps a names attribute ("Version"), which expect_identical
  # treats as a difference -- unname before comparing.
  got <- unname(trimws(read.dcf(file.path(datalib_test_pkg_root(), "DESCRIPTION"),
                                fields = "Version")[1, 1]))
  expect_identical(got, want)
})
